File: backend/models/found_item.py
"""ORM model for found items."""

from sqlalchemy import Column, String, Integer, DateTime, Text, JSON
from datetime import datetime
from database import Base


class FoundItem(Base):
    """Found item database record.

    Stores item details, registering municipality, and pickup conditions.
    """

    __tablename__ = "found_items"

    id = Column(String, primary_key=True, index=True)

    # Municipality info
    municipality_name = Column(String, nullable=False, index=True)
    municipality_type = Column(String, nullable=False)
    municipality_email = Column(String, nullable=False)

    # Item info
    item_name = Column(String, nullable=False, index=True)
    item_category = Column(String, nullable=False, index=True)
    item_date = Column(String, nullable=False)
    item_location = Column(String, nullable=False)
    item_status = Column(String, nullable=False, default="available")
    item_description = Column(Text, nullable=True)

    # Pickup info
    pickup_deadline = Column(Integer, nullable=False)
    pickup_location = Column(String, nullable=False)
    pickup_hours = Column(String, nullable=True)
    pickup_contact = Column(String, nullable=True)

    # Additional fields
    categories = Column(JSON, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, index=True)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    def to_dict(self):
        """Convert record to a dictionary matching the API schema."""
        return {
            "id": self.id,
            "municipality": {
                "name": self.municipality_name,
                "type": self.municipality_type,
                "contactEmail": self.municipality_email
            },
            "item": {
                "name": self.item_name,
                "category": self.item_category,
                "date": self.item_date,
                "location": self.item_location,
                "status": self.item_status,
                "description": self.item_description
            },
            "pickup": {
                "deadline": self.pickup_deadline,
                "location": self.pickup_location,
                "hours": self.pickup_hours,
                "contact": self.pickup_contact
            },
            "categories": self.categories or [],
            "createdAt": self.created_at.isoformat() if self.created_at else None,
            "updatedAt": self.updated_at.isoformat() if self.updated_at else None
        }

File: internal/model/found_item.go
package model

import (
	"database/sql"
	"encoding/json"
	"time"
)

type FoundItem struct {
	ID                string
	MunicipalityName  string
	MunicipalityType  string
	MunicipalityEmail string
	ItemName          string
	ItemCategory      string
	ItemDate          string
	ItemLocation      string
	ItemStatus        string
	ItemDescription   sql.NullString
	PickupDeadline    int
	PickupLocation    string
	PickupHours       sql.NullString
	PickupContact     sql.NullString
	Categories        sql.NullString
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

type MunicipalityInfo struct {
	Name         string `json:"name"`
	Type         string `json:"type"`
	ContactEmail string `json:"contactEmail"`
}

type ItemInfo struct {
	Name        string `json:"name"`
	Category    string `json:"category"`
	Date        string `json:"date"`
	Location    string `json:"location"`
	Status      string `json:"status"`
	Description string `json:"description,omitempty"`
}

type PickupInfo struct {
	Deadline int    `json:"deadline"`
	Location string `json:"location"`
	Hours    string `json:"hours,omitempty"`
	Contact  string `json:"contact,omitempty"`
}

type FoundItemCreate struct {
	Municipality MunicipalityInfo `json:"municipality"`
	Item         ItemInfo         `json:"item"`
	Pickup       PickupInfo       `json:"pickup"`
	Categories   []string         `json:"categories,omitempty"`
}

type FoundItemUpdate struct {
	Municipality *MunicipalityInfo `json:"municipality,omitempty"`
	Item         *ItemInfo         `json:"item,omitempty"`
	Pickup       *PickupInfo       `json:"pickup,omitempty"`
	Categories   *[]string         `json:"categories,omitempty"`
}

type FoundItemResponse struct {
	ID           string           `json:"id"`
	Municipality MunicipalityInfo `json:"municipality"`
	Item         ItemInfo         `json:"item"`
	Pickup       PickupInfo       `json:"pickup"`
	Categories   []string         `json:"categories"`
	CreatedAt    string           `json:"createdAt,omitempty"`
	UpdatedAt    string           `json:"updatedAt,omitempty"`
}

func (fi *FoundItem) ToResponse() FoundItemResponse {
	var cats []string
	if fi.Categories.Valid && fi.Categories.String != "" {
		_ = json.Unmarshal([]byte(fi.Categories.String), &cats)
	}
	if cats == nil {
		cats = []string{}
	}

	return FoundItemResponse{
		ID: fi.ID,
		Municipality: MunicipalityInfo{
			Name:         fi.MunicipalityName,
			Type:         fi.MunicipalityType,
			ContactEmail: fi.MunicipalityEmail,
		},
		Item: ItemInfo{
			Name:        fi.ItemName,
			Category:    fi.ItemCategory,
			Date:        fi.ItemDate,
			Location:    fi.ItemLocation,
			Status:      fi.ItemStatus,
			Description: fi.ItemDescription.String,
		},
		Pickup: PickupInfo{
			Deadline: fi.PickupDeadline,
			Location: fi.PickupLocation,
			Hours:    fi.PickupHours.String,
			Contact:  fi.PickupContact.String,
		},
		Categories: cats,
		CreatedAt:  fi.CreatedAt.Format(time.RFC3339),
		UpdatedAt:  fi.UpdatedAt.Format(time.RFC3339),
	}
}

type StatsResponse struct {
	FoundItems        FoundItemStats      `json:"foundItems"`
	TopCategories     []CategoryCount     `json:"topCategories"`
	TopMunicipalities []MunicipalityCount `json:"topMunicipalities"`
}

type FoundItemStats struct {
	Total     int `json:"total"`
	Available int `json:"available"`
	Claimed   int `json:"claimed"`
}

type CategoryCount struct {
	Category string `json:"category"`
	Count    int    `json:"count"`
}

type MunicipalityCount struct {
	Name  string `json:"name"`
	Count int    `json:"count"`
}

type ListParams struct {
	Skip         int
	Limit        int
	Category     string
	Municipality string
	Status       string
	Search       string
}
